js_modules/dagster-ui/packages/ui-core/src/overview/assets/AssetTimelineDataFetcher.tsx

```diff
@@ -0,0 +1,414 @@
+              assets.forEach((asset) => {
+                const stringKey = tokenForAssetKey(asset.key);
+                assetToGroup[stringKey] = groupId;
+              });
+              return assets;
+            })
+            .map((asset) => ({path: asset.key.path}));
+
+          if (!assetKeys.length) {
+            return Promise.resolve();
+          }
+
+          function triggerListeners() {
+            const listener = codeLocationSubscriptions.current[locationName];
+            if (listener) {
+              listener(runsCache.current.locations[locationName]!);
+            }
+
+            Object.keys(groups!).forEach((groupKey) => {
+              const listener = groupSubscriptions.current[groupKey];
+              if (listener) {
+                listener(runsCache.current.groups[groupKey]!);
+              }
+              assetKeys.forEach((key) => {
+                const stringAssetKey = tokenForAssetKey(key);
+                const listener = assetSubscriptions.current[stringAssetKey];
+                if (listener) {
+                  listener(runsCache.current.assets[stringAssetKey]!);
+                }
+              });
+            });
+          }
+
+          if (isRefresh) {
+            // Mark all of the assets as loading again
+            runsCache.current.locations[locationName] = {
+              runs: runsCache.current.locations[locationName]?.runs ?? [],
+              loading: true,
+            };
+            Object.keys(groups).forEach((groupKey) => {
+              runsCache.current.groups[groupKey] = {
+                runs: runsCache.current.groups[groupKey]?.runs ?? [],
+                loading: true,
+              };
+              assetKeys.forEach((key) => {
+                const stringAssetKey = tokenForAssetKey(key);
+                runsCache.current.assets[stringAssetKey] = {
+                  runs: runsCache.current.assets[stringAssetKey]?.runs ?? [],
+                  loading: true,
+                };
+              });
+            });
+            triggerListeners();
+          }
+
+          return new Promise(async (res) => {
+            const {data} = await client.query<AssetTimelineQuery, AssetTimelineQueryVariables>({
+              query: ASSET_TIMELINE_QUERY,
+              variables: {
+                inProgressFilter: {
+                  statuses: [RunStatus.CANCELING, RunStatus.STARTED],
+                  createdBefore: endSecRef.current / 1000,
+                  assetKeys,
+                },
+                terminatedFilter: {
+                  statuses: Array.from(doneStatuses),
+                  createdBefore: endSecRef.current / 1000,
+                  updatedAfter: startSecRef.current / 1000,
+                  assetKeys,
+                },
+                tickCursor: startSecRef.current / 1000,
+                ticksUntil: endSecRef.current / 1000,
+              },
+            });
+            res(void 0);
```

Comment:
It's odd that this promise returns before it finishes executing. I think this is worthy of a comment block!

Reply:
Might also be nice to kick this client.query or the `variables` construction to a little helper method to get it out of the way fo the logic too